### PR TITLE
Baikal-t qemu

### PR DESCRIPTION
**Pull request by anton-bondarev**
* Add mips p5600 gic interrupt controller
* Improve mips arch
* Enable ethernet in baikal-t on qemu

### CODE REVIEW

File: src/drivers/interrupt/mips_gic/mips_gic.h [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+#define __IRQCTRL_IRQS_TOTAL      192
+
+#define MIPS_GIC_INTERRUPT_PIN    3
+
+#define GIC_SH_CONFIG           0x0000
+#define GIC_SH_COUNTER_LO       0x0010
+#define GIC_SH_COUNTER_HI       0x0014
+#define GIC_SH_REVID            0x0020
+
+#define GIC_SH_INT_AVAIL(N)    (0x0024 + ((N >> 5) << 2))
```
Comment (alexkalmuk) [inline src/drivers/interrupt/mips_gic/mips_gic.h]: It's safer to use `(N)` here and below.

File: src/arch/mips/include/asm/mipsregs.h [hunk 2]
```diff
@@ -26,6 +26,12 @@
 #define CP0_EBASE    15    /* select 1 exception base */
 #define CP0_CONFIG   16    /* select 0 Configuration */
 #define CP0_CONFIG1  16    /* select 1 configuration */
+#define CP0_CONFIG2  16    /* select 2 Configuration */
+#define CP0_CONFIG3  16    /* select 3 configuration */
+#define CP0_CONFIG4  16    /* select 4 configuration */
+#define CP0_CONFIG5  16    /* select 5 Configuration */
+#define CP0_CONFIG6  16    /* select 6 configuration */
+#define CP0_CONFIG7  16    /* select 7 configuration */
```
[the remark this answers is not in the document]
Comment (anton-bondarev) [reply src/arch/mips/include/asm/mipsregs.h]: Yes, they are all registers 16 but different selects

[the remark this answers is not in the document]
Comment (alexkalmuk) [reply src/arch/mips/include/asm/mipsregs.h]: Ok, thanks.

File: src/arch/mips/kernel/interrupt.c [hunk 3]
```diff
@@ -16,22 +18,70 @@
...
+#include <stdint.h>
+
+#ifdef MIPS_GIC_INTERRUPT_PIN
+void mips_gic_interrupt(void) {
+	uint32_t pend;
+	int i;
+
+	for (i = 0; i < __IRQCTRL_IRQS_TOTAL; i += 32) {
+		pend = REG32_LOAD(0x1BDC0000 + GIC_SH_PEND(i)) &
+				REG32_LOAD(0x1BDC0000 + GIC_SH_MASK(i));
```
[the remark this answers is not in the document]
Comment (anton-bondarev) [reply src/arch/mips/kernel/interrupt.c]: done

File: src/arch/mips/subarch/p5600/mips32_p5600_arch.c [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+void arch_init(void) {
+	uint32_t status;
+	uint32_t cause;
+	uint32_t intctl;
+	uint32_t config3;
+
+	status = mips_read_c0_status();
+	if (status & ST0_BEV) {
+
+	}
```
[the remark this answers is not in the document]
Comment (anton-bondarev) [reply src/arch/mips/subarch/p5600/mips32_p5600_arch.c]: done

File: src/drivers/interrupt/mips_gic/mips_gic.c [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	int i;
+
+	c0 = mips_read_c0_status();
+	c0 |= 1 << (MIPS_GIC_INTERRUPT_PIN + ST0_IRQ_MASK_OFFSET + ST0_SOFTIRQ_NUM);
+	mips_write_c0_status(c0);
+
+	mips32_gcb_set_register(GCR_GIC_BASE, MIPS_GIC_BASE | GIC_EN);
+	log_info("mips_gic revision %x", REG_LOAD(MIPS_GIC_BASE + GIC_SH_REVID));
+
+	for (i = 0; i < 256; i++) {
```
[the remark this answers is not in the document]
Comment (anton-bondarev) [reply src/drivers/interrupt/mips_gic/mips_gic.c]: done

File: src/drivers/interrupt/mips_gic/mips_gic.c [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+		c0 = mips_read_c0_status();
+		c0 |= 1 << (interrupt_nr + ST0_IRQ_MASK_OFFSET);
+		mips_write_c0_status(c0);
+		return;
+	}
+
+	reg = interrupt_nr >> 5;
+	mask = (1 << (interrupt_nr - (reg << 5)));
+
+	REG32_STORE(MIPS_GIC_BASE + GIC_SH_SMASK(reg << 5), mask);
```
[the remark this answers is not in the document]
Comment (anton-bondarev) [reply src/drivers/interrupt/mips_gic/mips_gic.c]: done

File: src/drivers/interrupt/mips_gic/mips_gic.c [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+		REG32_STORE(MIPS_GIC_BASE + GIC_SH_MAP_PIN(i), MIPS_GIC_INTERRUPT_PIN | 1 << 31);
+	}
+	return 0;
+}
+
+void irqctrl_enable(unsigned int interrupt_nr) {
+	int reg;
+	uint32_t mask;
+
+	if (interrupt_nr > __IRQCTRL_IRQS_TOTAL) {
```
[the remark this answers is not in the document]
Comment (anton-bondarev) [reply src/drivers/interrupt/mips_gic/mips_gic.c]: done

File: src/drivers/interrupt/mips_gic/mips_gic.h (hunk 1 above)
Comment (anton-bondarev) [reply src/drivers/interrupt/mips_gic/mips_gic.h]: done

**Pull request merged**